src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -0,0 +1,216 @@
+            while (baseType != null)
+            {
+                if (GeneratorHelper.IsWinRTType(baseType))
+                {
+                    compatibleTypesForGeneric.Add(baseType);
+                }
+                baseType = baseType.BaseType;
+            }
+
+            typeStack.Pop();
+
+            compatibleTypes = new List<INamedTypeSymbol>(compatibleTypesForGeneric.Count);
+            foreach (var compatibleType in compatibleTypesForGeneric)
+            {
+                compatibleTypes.Add(definition.Construct(compatibleType));
+            }
+
+            return true;
+        }
+
+        private static void GenerateVtableAttributes(SourceProductionContext sourceProductionContext, VtableAttribute vtableAttribute)
+        {
+            // Even if no interfaces are on the vtable, as long as the base type is a WinRT type,
+            // we want to put the attribute as the attribute is the opt-in for the source generated
+            // vtable generation.
+            if (vtableAttribute.Interfaces.Any() || vtableAttribute.HasWinRTExposedBaseType)
+            {
+                var vtableEntries = string.Join(", ", vtableAttribute.Interfaces.Select((@interface) => $@"typeof({@interface})"));
+                
+                StringBuilder source = new();
+                if (!vtableAttribute.IsGlobalNamespace)
+                {
+                    source.Append($$"""
+                        namespace {{vtableAttribute.Namespace}}
+                        {
+                        """);
+                }
+
+                source.Append($$"""
+                    [global::WinRT.WinRTExposedType({{vtableEntries}})]
+                    partial class {{vtableAttribute.ClassName}}
+                    {
+                    }
+                    """);
+
+                if (!vtableAttribute.IsGlobalNamespace)
+                {
+                    source.Append($@"}}");
+                }
+
+                string prefix = vtableAttribute.IsGlobalNamespace ? "" : $"{vtableAttribute.Namespace}.";
+                sourceProductionContext.AddSource($"{prefix}{vtableAttribute.ClassName}.WinRTVtable.g.cs", source.ToString());
+            }
+        }
+    }
+
+    internal sealed record VtableAttribute
+    {
+        public string Namespace { get; }
+        public bool IsGlobalNamespace { get; }
+        public string ClassName { get; }
+        public EquatableArray<string> Interfaces { get; }
+        public bool HasWinRTExposedBaseType { get; }
+
+        public VtableAttribute(string @namespace, bool isGlobalNamespace, string className, EquatableArray<string> interfaces, bool hasWinRTExposedBaseType)
+        {
+            Namespace = @namespace;
+            IsGlobalNamespace = isGlobalNamespace;
+            ClassName = className;
+            Interfaces = interfaces;
+            HasWinRTExposedBaseType = hasWinRTExposedBaseType;
+        }
+    }

```

Comment:
Nit: now that you have PolySharo, you can just use the record syntax to keep it more compact 🙂
```suggestion
    internal sealed record VtableAttribute(
        string Namespace,
        bool IsGlobalNamespace,
        string ClassName,
        EquatableArray<string> Interfaces,
        bool HasWinRTExposedBaseType);
```